backend/src/main/java/com/google/sps/servlets/AddParticipantServlet.java

```diff
@@ -66,109 +49,96 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
     Participant newParticipant =
         new Participant(
-            /* id= */ -1L, username, startTimeAvailable, endTimeAvailable, duration, timestamp);
-
+            username,
+            startTimeAvailable,
+            endTimeAvailable,
+            duration,
+            /* currentMatchId=*/ 0,
+            timestamp);
+
+    // Get DatastoreService and instiate Match and Participant Datastores
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    MatchDatastore matchDatastore = new MatchDatastore(datastore);
+    ParticipantDatastore participantDatastore = new ParticipantDatastore(datastore);
 
-    // Find immediate match if possibl;
-    FindMatchQuery query = new FindMatchQuery(Clock.systemUTC());
-    Match match = query.findMatch(getParticipants(datastore), newParticipant);
+    // Find immediate match if possible
+    FindMatchQuery query = new FindMatchQuery(Clock.systemUTC(), participantDatastore);
+    Match match = query.findMatch(newParticipant);
 
-    // Match found, add to datastore, delete matched participants from datastore
     if (match != null) {
-      addMatchToDatastore(match, datastore);
-      deleteParticipantFromDatastore(match.getSecondParticipant(), datastore);
+      // Match found, add to match datastore, update participant datastore
+      long matchId = matchDatastore.addMatch(match);
+
+      // Update currParticipant entity with new currentMatchId and null availability
+      Participant currParticipant =
+          participantDatastore.getParticipantFromUsername(match.getSecondParticipantUsername());
+      currParticipant.setCurrentMatchId(matchId);
+      currParticipant.setStartTimeAvailable(null);
+      currParticipant.setEndTimeAvailable(null);
+      currParticipant.setDuration(0);
+      participantDatastore.addParticipant(currParticipant);
+
+      // Update newParticipant entity with new currentMatchId and null availability
+      newParticipant.setCurrentMatchId(matchId);
+      newParticipant.setStartTimeAvailable(null);
+      newParticipant.setEndTimeAvailable(null);
+      newParticipant.setDuration(0);
+      participantDatastore.addParticipant(newParticipant);
     } else {
-      // Match not found, insert participant entity into datastore
-      Entity participantEntity = new Entity(KEY_PARTICIPANT);
-      participantEntity.setProperty(PROPERTY_USERNAME, username);
-      participantEntity.setProperty(PROPERTY_STARTTIMEAVAILABLE, startTimeAvailable);
-      participantEntity.setProperty(PROPERTY_ENDTIMEAVAILABLE, endTimeAvailable);
-      participantEntity.setProperty(PROPERTY_DURATION, duration);
-      participantEntity.setProperty(PROPERTY_TIMESTAMP, timestamp);
-      datastore.put(participantEntity);
+      // Match not found, add participant to datastore
+      participantDatastore.addParticipant(newParticipant);
     }
 
     // Redirect back to the HTML page
     response.sendRedirect("/index.html");
```

Comment:
Since we are checking if there is an immediate match or not, I think we should also return the resulting match status to the frontend

Reply:
When an immediate match is found, the participant doesn't have to be immediately notified since the javascript part will search for the match every 5 seconds anyways.

Reply:
I feel like if we are searching for a match immediately then it might be worth sending back the resulting match immediately as well instead of waiting to return and traversing a list of all matches and then returning it